pandas/io/parsers/arrow_parser_wrapper.py

```diff
@@ -139,6 +141,30 @@ def handle_warning(invalid_row) -> str:
                 f"f{n}" for n in self.convert_options["include_columns"]
             ]
 
+        if self.dtype is not None:
+            if isinstance(self.dtype, dict):
+                column_types = {}
+                for col, col_dtype in self.dtype.items():
+                    source_dtype = pandas_dtype(col_dtype)
+
+                    try:
+                        target_dtype = to_pyarrow_type(source_dtype.type)
+                        if target_dtype:
+                            column_types[col] = target_dtype
+
+                    except TypeError:
+                        # TODO: Unsupported dtypes silently ignored - may cause
+                        # unexpected behavior when pyarrow applies default inference
+                        # instead of user's dtype
+                        pass
+
+                if column_types:
+                    self.convert_options["column_types"] = column_types
+            else:
+                # TODO: Global dtypes not supported - may cause inconsistent behavior
+                # between engines, especially for leading zero preservation
+                pass
```

Comment:
if they pass a singleton, can we do something like

`convert_options["column_types"] = defaultdict(user_passed_dtype)`

?

Reply:
It's a good thought, but it doesn't work. One of the first things I tried actually :) I documented a larger analysis on pyarrow here: https://github.com/apache/arrow/issues/47502

You can see the relevant portion of pyarrow code here. Everything is mapped back to C++, and if the column name is not found, it uses the default option.

https://github.com/apache/arrow/blob/eb9d5194a306f8145f8600b176f3bd391ee4397c/cpp/src/arrow/csv/reader.cc#L675-L682